---
title: "Master script for TNBC typing project"
author: "Amy Gill"
date: "5/4/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script combines several independent scripts so that all the code for this project is in one place.

Component scripts:

`1-get-data.Rmd`


## Libraries and color scheme

```{r message=FALSE}
library(stringr)
library(tidyverse)
library(SingleCellExperiment)
library(Seurat)
library(pdftools)

colorblind_palette <- c("#000000", "#D55E00",  "#56B4E9", "#E69F00", "#009E73", "#F0E442", "#0072B2",  "#CC79A7")
```

## Import single cell data from GEO

### Original paper

Karaayvaz M, Cristea S, Gillespie SM, Patel AP et al. Unravelling subclonal heterogeneity and aggressive disease states in TNBC through single-cell RNA-seq. Nat Commun 2018 Sep 4;9(1):3588. PMID: 30181541

https://www-ncbi-nlm-nih-gov.proxy1.library.jhu.edu/geo/query/acc.cgi?acc=GSE118389

### Dataset on GEO

GEO link: https://www-ncbi-nlm-nih-gov.proxy1.library.jhu.edu/geo/query/acc.cgi?acc=GSE118389

### Download data

#### TPM counts before normalization

- Downloaded `GSE118389_tpm_rsem.txt.gz` to `data/`.
- Unzipped with `gunzip data/GSE118389_tpm_rsem.txt.gz`

#### Normalized data

- Downloaded `GSE118398_norm_data.txt.gz` to `data/`.
- Unzipped with `gunzip data/GSE118389_norm_data.txt.gz`

#### Metadata

- Downloaded `GSE118389_series_matrix.txt.gz` to `data/`.
- Unzipped with `gunzip data/GSE118389_series_matrix.txt.gz`

##### Overall design

"Single cell RNA sequencing of 1,534 cells in six fresh triple negative breast cancer tumors."

##### Sample collection protocol

"Fresly collected human tumor samples were mechanically and enzymatically digested using tumor dissociation kit (Miltenyi Biotec). Single cells were sorted by flow cytometry, lysed and snap frozen on dry ice immediately."

##### Library preparation

"RNA libraries were prepared using Smart-seq2 protocol (Picelli et al., 2014)"

##### Sequencing instrument

Illumina Genome Analyzer

##### Genome build

GRCh38

##### TPM calculations

"Fastq files were quantified to transcript per million (TPM) expression values with RSEM with default parameters"

##### Removing low quality cells

"Low quality cells were filtered as follows: 4 median absolute deviations (MADs) below the median for any of the three metrics concering: i) library size; ii) number of expressed genes; iii) total amount of mRNA"

##### Removing unexpressed genes

"Unexpressed genes were removed as follows: a gene was considered unexpressed if log2(TPM+1) < 0.1 in at least 95% of cells separately for each patient"

##### Normalization

"Data was normalized following a three-step combination approach: i) Transform the TPM values into relative counts with the Census algorithm (Qiu et al, 2017); ii) Normalize the Census counts with the deconvolution strategy implemented in scran (Lun et al 2016); iii) Remove additional sources of variation with RUVSeq (Risso et al, 2014)"



### Format normalized data

#### Read normalized data

```{r}
# import as data frame
tnbc <- read.table("data/GSE118389_norm_data.txt")

# convert to matrix
tnbc_mat <- as.matrix(tnbc)
```

#### Create column data

All samples are from human triple negative breast cancer and were processed identically. The main factor that differs between samples is that they come from 6 different patients. Patient ID is embedded in the sample name.

```{r}
# extract patient ID from column name
patient_id <- sapply(1:length(colnames(tnbc)), function(i){
	str_split(colnames(tnbc)[i], pattern = "_")[[1]][1]    # portion of the column name before the first underscore
})

# table of number of cells per patient - as expected, matches original publication
data.frame(table(patient_id)) %>%
	mutate("Patient ID" = patient_id,
		   "Number of cells" = Freq) %>%
	select(-c(patient_id, Freq))
```

```{r}
# bind sample ID and patient ID into data frame
coldata <- DataFrame(sample_id = colnames(tnbc),
					  patient_id = patient_id)

# set row names to sample ID for compatibility with SingleCellExperiment format
rownames(coldata) <- colnames(tnbc)
```

#### Create row data

Initial row data consists only of gene names, which are currently the row names of the expression matrix

```{r}
# add feature symbols (gene names)
rowdata <- data.frame(feature_symbol = rownames(tnbc))

# add row names for proper SingleCellExperiment functionality (gene names)
rownames(rowdata) <- rownames(tnbc)
```

#### Convert to SingleCellExperiment format

```{r}
sce <- SingleCellExperiment(tnbc_mat)    # add normalized count data
colData(sce) <- coldata    # add sample info
rowData(sce) <- rowdata    # add gene info

# set name of assay for compatibility with downstream tools
# (some algorithms threw errors when there was no counts or logcounts, and these data are not log-normalized)
names(assays(sce)) <- "normcounts"

# save to rda format
save(sce, file = "rdata/sce.rda")
```

### Create alternate SingleCellExperiment with TPM counts before normalization

SC3 requires the TPM counts in addition to the normalized counts in order to calculate dropout. I created a separate `sce_tpm` object for this instead of modifying `sce` because I had already done downstream processing that used the normcounts slot, and I didn't want to risk breaking anything.

```{r}
# read TPM data
tpm <- read.table("data/GSE118389_tpm_rsem.txt")
tpm <- as.matrix(tpm)

# extract patient ID from column names
pt_id_tpm <- sapply(1:length(colnames(tpm)), function(i){
	str_split(colnames(tpm)[i], pattern = "_")[[1]][1]    # portion of the column name before the first underscore
})

# bind sample ID and patient ID into data frame
coldata_tpm <- DataFrame(sample_id = colnames(tpm),
					  patient_id = pt_id_tpm)

# set row names to sample ID for compatibility with SingleCellExperiment format
rownames(coldata_tpm) <- colnames(tpm)

# create rowData from row names
rowdata_tpm <- data.frame(feature_symbol = rownames(tpm))
rownames(rowdata_tpm) <- rownames(tpm)

# create SingleCellExperiment with TPM
sce_tpm <- SingleCellExperiment(tpm)    # add TPM data
colData(sce_tpm) <- coldata_tpm    # add sample info
rowData(sce_tpm) <- rowdata_tpm    # add gene info

names(assays(sce_tpm)) <- "counts"

# save as rda
save(sce_tpm, file = "rdata/sce_tpm.rda")
```

### Create Seurat object from SingleCellExperiment

```{r}
tnbc_seurat <- as.Seurat(sce, counts = "normcounts", data = "normcounts")
```
### Checkpoint: Seurat object before scaling and regression

```{r}
tnbc_seurat_pre_reg <- tnbc_seurat
save(tnbc_seurat_pre_reg, file = "rdata/tnbc_seurat_pre_reg.rda")
```


## Preprocessing

### Regress out patient effect

In the original paper, clustering was performed after regressing out patient-specific effects. Therefore, I will do the same.

```{r}
# scale data while regressing out patient_id variable
tnbc_seurat <- ScaleData(tnbc_seurat, vars.to.regress = "patient_id", features = rownames(tnbc_seurat))
```

### Find variable features

```{r}
# identify 2000 most variable features
tnbc_seurat <- FindVariableFeatures(tnbc_seurat, selection.method = "vst", nfeatures = 2000)

# create variable feature plot
p1 <- VariableFeaturePlot(tnbc_seurat)

# label top 8 most variable features
top8 <- head(VariableFeatures(tnbc_seurat), 8)
p2 <- LabelPoints(plot = p1, points = top8, repel = TRUE, xnudge = 0.05, ynudge = 0.01)

# view variable feature plot
p2

# save labeled variable feature plot
ggsave(p2, filename = "results/var-features.png", height = 4, width = 7)
```

### Dimensionality reduction and projection

#### Principal component analysis

```{r}
# perform PCA with the 2000 most variable features
tnbc_seurat <- RunPCA(tnbc_seurat, features = VariableFeatures(tnbc_seurat))
```

#### tSNE

```{r}
# run tSNE on the top 50 PCs
tnbc_seurat <- RunTSNE(tnbc_seurat, dims = 1:50)
```


#### UMAP

```{r}
# run UMAP on the top 50 PCs with the R-native UWOT method and cosine metric
tnbc_seurat <- RunUMAP(tnbc_seurat, dims = 1:50)
```

### Color tSNE and UMAP by patient ID

#### tSNE

```{r}
# color tSNE plot by patient ID
patient_tsne <- DimPlot(tnbc_seurat, reduction = "tsne", group.by = "patient_id") +
	ggtitle("tSNE plot of patients after removing patient effect") +
	scale_color_manual(values = colorblind_palette[c(1:5,8)])

# inspect tSNE plot colored by patient ID
patient_tsne

# save tSNE plot colored by patient ID
ggsave(patient_tsne, filename = "results/patient-tsne.png", width = 7, height = 4)
```

#### UMAP

```{r}
# color UMAP plot by patient ID
patient_umap <- DimPlot(tnbc_seurat, reduction = "umap", group.by = "patient_id") +
	ggtitle("UMAP plot of patients after removing patient effect") +
	scale_color_manual(values = colorblind_palette[c(1:5,8)])

# inspect UMAP plot colored by patient ID
patient_umap

# save UMAP plot colored by patient ID
ggsave(patient_umap, filename = "results/patient-umap.png", width = 7, height = 4)
```


## Louvain clustering with Seurat

### Find neighbors

```{r}
# compute nearest neighbor graph and shared nearest neighbors (SNN) at the cell level
tnbc_seurat <- FindNeighbors(tnbc_seurat, dims = 1:50)
```

### Testing clustering resolutions

The tested range of resolution parameters was suggested by the [Seurat clustering tutorial](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html) for datasets of up to 3000 cells.

```{r}
# run clustering with a range of resolution parameters
rez <- seq(0.4, 1.2, 0.1)
clust <- sapply(rez, function(x){
	trez <- FindClusters(tnbc_seurat, resolution = x)
	length(table(Idents(trez)))
})
rez_clust <- data.frame(rez, clust)

# scatter plot and curve fit for number of clusters versus Louvain resolution parameter
gg_rez_clust <- rez_clust %>%
	ggplot(aes(rez, clust)) +
	geom_point() + 
	geom_line() +
	xlab("Resolution parameter of Louvain clustering") +
	ylab("Number of clusters") +
	ggtitle("Clusters identified at different resolutions by the Louvain algorithm") +
	ylim(c(5,20))

# view plot of number of clusters versus resolution
gg_rez_clust

# save clustering resolution plot
ggsave(gg_rez_clust, file = "results/clustering-resolutions.png", height = 4, width = 6)
```

At resolutions greater than or equal to 1, there are 18 clusters. Moving forward, I use a resolution of 1.

### Perform clustering

```{r}
# Louvain clustering with a resolution of 1
tnbc_seurat <- FindClusters(tnbc_seurat, resolution = 1)

# number of cells per Louvain cluster
table(tnbc_seurat$seurat_clusters)
```

### tSNE plot of Louvain clusters

```{r}
# color tSNE plot by Louvain clusters
louvain_tsne <- DimPlot(tnbc_seurat, reduction = "tsne",
		group.by = "seurat_clusters",
		label = TRUE, label.size = 5, repel = TRUE) +
	ggtitle("tSNE plot of Louvain clusters")

# inspect tSNE plot of Louvain clusters
louvain_tsne

# save tSNE plot of Louvain clusters
ggsave(louvain_tsne, filename = "results/louvain_tsne.png", height = 4, width = 7)
```


### UMAP plot of Louvain clusters

```{r}
# color UMAP plot by Louvain clusters
louvain_umap <- DimPlot(tnbc_seurat, reduction = "umap",
		group.by = "seurat_clusters",
		label = TRUE, label.size = 5, repel = TRUE) +
	ggtitle("UMAP plot of Louvain clusters")

# inspect UMAP plot of Louvain clusters
louvain_umap

# save UMAP plot of Louvain clusters
ggsave(louvain_umap, filename = "results/louvain_umap.png", height = 4, width = 7)
```
### Differentially expressed genes in Seurat clusters

```{r}
# find genes differentially expressed in one cluster versus all other clusters with log fold change >= 0.5
seurat_markers <- FindAllMarkers(tnbc_seurat, logfc.threshold = 0.5, slot = "scale.data")

# keep only differentially expressed markers with an adjusted pval less than .05
seurat_markers <- seurat_markers %>%
	filter(p_val_adj <= .05)

# save marker genes
save(seurat_markers, file = "rdata/seurat_markers.rda")

# define top 5 differentially expressed genes in each cluster
top5 <- seurat_markers %>%
	group_by(cluster) %>%
	top_n(5, wt = avg_diff)

louvain_de_heat <- DoHeatmap(tnbc_seurat, features = top5$gene) + NoLegend()

louvain_de_heat

ggsave(louvain_de_heat, filename = "results/louvain-de-genes-heat.png", width = 10, height = 10)
```

### Save updated Seurat and SingleCellExperiment objects

```{r}
# save updated tnbc_seurat object
save(tnbc_seurat, file = "rdata/tnbc_seurat.rda")

# create updated sce object
sce_post_seurat <- as.SingleCellExperiment(tnbc_seurat)
save(sce_post_seurat, file = "rdata/sce_post_seurat.rda")
```

## Aside: tSNE and UMAP showing patient ID before regressing out patient effect

### Preprocessing to run tSNE and UMAP before removing patient effect

```{r}
# find variable features without removing patient effect
tnbc_seurat_pre_reg <- FindVariableFeatures(tnbc_seurat_pre_reg, selection.method = "vst", nfeatures = 2000)

# center and scale data without removing patient effect
tnbc_seurat_pre_reg <- ScaleData(tnbc_seurat_pre_reg, features = rownames(tnbc_seurat_pre_reg))

# PCA without removing patient effect
tnbc_seurat_pre_reg <- RunPCA(tnbc_seurat_pre_reg, features = VariableFeatures(tnbc_seurat_pre_reg))
```

### tSNE plot of patients before removing patient effect

```{r}
# run tSNE without removing patient effect
tnbc_seurat_pre_reg <- RunTSNE(tnbc_seurat_pre_reg, dims = 1:50)

# create tSNE plot without removing patient effect and color by patient
tsne_patient_pre_reg <- DimPlot(tnbc_seurat_pre_reg, reduction = "tsne",
		group.by = "patient_id") +
	ggtitle("tSNE plot of patients before removing patient effect") +
	scale_color_manual(values = colorblind_palette[c(1:5,8)])

# inspect tSNE plot created without removing patient effect
tsne_patient_pre_reg

# save tSNE plot created without removing patient effect
ggsave(tsne_patient_pre_reg, filename = "results/tsne-patient-pre-reg.png", width = 7, height = 4)
```

### UMAP plot of patients before removing patient effect

```{r}
# run UMAP without removing patient effect
tnbc_seurat_pre_reg <- RunUMAP(tnbc_seurat_pre_reg, dims = 1:50)

# create tSNE plot without removing patient effect and color by patient
umap_patient_pre_reg <- DimPlot(tnbc_seurat_pre_reg, reduction = "umap",
		group.by = "patient_id") +
	ggtitle("UMAP plot of patients before removing patient effect") +
	scale_color_manual(values = colorblind_palette[c(1:5,8)])

# inspect UMAP plot created without removing patient effect
umap_patient_pre_reg

# save tSNE plot created without removing patient effect
ggsave(umap_patient_pre_reg, filename = "results/umap-patient-pre-reg.png", width = 7, height = 4)
```

## Marker gene based cell typing (a.k.a literature-based cell typing)

### Importing marker gene list

These data were imported from the PDF of supplementary table 7 of Karaayvaz et al., 2018.

#### Paper description of marker genes and classification strategy

"The literature-based list used to classify cell types consists of 49 expression markers specific to four cell types, compiled from multiple references (Tirosh et al., 2016) (Table 7). In order to minimize the number of misassignments, we only assign cells to a specific cell type when we consider that there is enough expression-based evidence supporting the assignment (expression threshold = 1). To this end, we derive a series of expert-based rules:

1. Epithelial class: a cell has epithelial characteristics if it expresses either:

- at least 2 epithelial markers;

- only one of the strongest breast epithelial markers: EPCAM, KRT8, KRT18, KRT19, with expression
higher than in 50% of the cells for that respective patient, for that marker.

2. Specific immune class: a cell has specific immune characteristics (T cell, B cell, macrophage) if it expresses
either:

-only specific immune markers of that type (at least 2);

-PTPRC and only specific immune markers of that type (at least 1);

-at least 3 immune markers of that type, and at most 1 immune marker of another type.

3. Stroma class: a cell has stroma characteristics if it expresses either:

- only stroma markers;

- at least 3 stroma markers and at most 1 endothelial marker.

4. Endothelial class: a cell has endothelial characteristics if it expresses either:

- only endothelial markers;

- at least 3 endothelial markers and at most 1 stroma marker."

#### Import marker gene table from PDF of supplement

```{r}
supp <- pdf_text("paper/supplement.pdf")
```

Supplementary table 7 is split across pages 35 and 36. First extract the rows of the table from page 35:

```{r}
# extract page 35 of the supplement as text
p35 <- supp[[35]]

# split into a vector of lines
p35_split <- str_split(p35, "\r\n")[[1]]

# remove leading and trailing whitespace
p35_split <- str_trim(p35_split)

# start building vector to turn into table 
s7_dat <- p35_split[14:50]
```

Append the rows of the table from page 36:

```{r}
# extract page 36 of the supplement as text
p36 <- supp[[36]]

# split into a vector of lines
p36_split <- str_split(p36, "\r\n")[[1]]

# remove leading and trailing whitespace
p36_split <- str_trim(p36_split)

# append to vector 
s7_dat <- c(s7_dat, p36_split[1:12])
```

Split the elements of the vector on whitespace. The first chunk of each element is the marker gene, the second chunk is the cell type indicated by that marker gene, and the third chunk (when present) is the cell subtype indicated by that marker gene.

```{r}
# split on more than one space
s7_split <- str_split(s7_dat, pattern = "\\s\\s+")

# extract marker gene name (first elements of s7_split)
marker <- sapply(1:length(s7_split), function(i){
	s7_split[[i]][1]
})

# extract cell type indicated by marker (second elements of s7_split)
cell_type <- sapply(1:length(s7_split), function(i){
	s7_split[[i]][2]
})

# extract cell subtype indicated by marker (third elements of s7_split, if they exist)
cell_subtype <- sapply(1:length(s7_split), function(i){
	ifelse(length(s7_split[[i]]) > 2, s7_split[[i]][3], NA)
})
```

Combine the marker gene, cell type and cell subtype into a data frame, then save that data frame.

```{r}
# create data frame of cell type markers
cell_type_markers <- data.frame(marker = marker,
		   cell_type = cell_type,
		   cell_subtype = cell_subtype)

# save table of cell type markers
save(cell_type_markers, file = "rdata/cell_type_markers.rda")
```

